crates/typst-library/src/model/par.rs

```diff
@@ -108,6 +108,132 @@ pub struct ParElem {
     /// to `{-0.2em}` to get a baseline gap of exactly `{2em}`. The exact
     /// distribution of the top- and bottom-edge values affects the bounds of
     /// the first and last line.
+    ///
+    /// ```preview
+    /// // Color palette
+    /// #let c = (
+    ///   par-line: aqua.transparentize(60%),
+    ///   leading-line: blue,
+    ///   leading-text: blue.darken(20%),
+    ///   spacing-line: orange,
+    ///   spacing-text: orange.darken(10%),
+    /// )
+    ///
+    /// // A sample text for measuring font metrics.
+    /// #let sample-text = [A]
+    ///
+    /// // The wide margin is for annotations
+    /// #set page(width: 350pt, margin: (x: 20%))
+    ///
+    /// #context {
+    ///   let text-height = measure(sample-text).height
+    ///   let line-height = text-height + par.leading.to-absolute()
+    ///
+    ///   // Number of lines in each paragraph
+    ///   let n-lines = {
+    ///     // Calculate `n-lines` by querying `par.line`.
+    ///     let positions = query(<par-line>).map(it => it.location().position().y)
+    ///
+    ///     // Determine it's leading or spacing
+    ///     let kinds = positions
+    ///       .enumerate()
+    ///       .slice(1)
+    ///       .map(((i, y)) => {
+    ///         // Displacement from the last line relative to the line height
+    ///         let jump = (positions.at(i) - positions.at(i - 1)) / line-height
+    ///         if calc.abs(jump - 1) < 0.1 { "leading" } else { "spacing" }
+    ///       })
+    ///
+    ///     kinds.split("spacing").map(x => x.len() + 1)
+    ///   }
```

Comment:
> But it's the 79 line context that might be shrinkable. 

Yes, I'm also worried about this.

For example, the whole `let n-lines = { … }` can be replaced with `let n-lines = (4, 4, 2)`.
Not sure which way is better…

Reply:
I think keeping it simple is better